ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnchainPrivacyController.java

```diff
@@ -0,0 +1,367 @@
+        privacyParameters.getPrivateStateStorage(),
+        privacyParameters.getEnclave(),
+        new PrivateTransactionValidator(chainId),
+        privateTransactionSimulator,
+        privateNonceProvider,
+        privateWorldStateReader,
+        privacyParameters.getPrivateStateRootResolver());
+  }
+
+  public OnchainPrivacyController(
+      final Blockchain blockchain,
+      final PrivateStateStorage privateStateStorage,
+      final Enclave enclave,
+      final PrivateTransactionValidator privateTransactionValidator,
+      final PrivateTransactionSimulator privateTransactionSimulator,
+      final PrivateNonceProvider privateNonceProvider,
+      final PrivateWorldStateReader privateWorldStateReader,
+      final PrivateStateRootResolver privateStateRootResolver) {
+    super(
+        blockchain,
+        privateStateStorage,
+        enclave,
+        privateTransactionValidator,
+        privateTransactionSimulator,
+        privateNonceProvider,
+        privateWorldStateReader,
+        privateStateRootResolver);
+
+    onchainPrivacyGroupContract = new OnchainPrivacyGroupContract(privateTransactionSimulator);
+  }
+
+  @Override
+  public String createPrivateMarkerTransactionPayload(
+      final PrivateTransaction privateTransaction,
+      final String privacyUserId,
+      final Optional<PrivacyGroup> privacyGroup) {
+    final String firstPart;
+    try {
+      LOG.trace("Storing private transaction in enclave");
+      final SendResponse sendResponse = sendRequest(privateTransaction, privacyGroup);
+      firstPart = sendResponse.getKey();
+    } catch (final Exception e) {
+      LOG.error("Failed to store private transaction in enclave", e);
+      throw e;
+    }
+    final Optional<String> optionalSecondPart =
+        buildAndSendAddPayload(
+            privateTransaction,
+            Bytes32.wrap(privateTransaction.getPrivacyGroupId().orElseThrow()),
+            privacyUserId);
+
+    return buildCompoundLookupId(firstPart, optionalSecondPart);
+  }
+
+  @Override
+  public Optional<PrivacyGroup> findPrivacyGroupByGroupId(
+      final String privacyGroupId, final String enclaveKey) {
+    // get the privateFor list from the management contract
+    final Optional<TransactionProcessingResult> privateTransactionSimulatorResultOptional =
+        privateTransactionSimulator.process(
+            privacyGroupId, buildCallParams(GET_PARTICIPANTS_METHOD_SIGNATURE));
+
+    if (privateTransactionSimulatorResultOptional.isPresent()
+        && privateTransactionSimulatorResultOptional.get().isSuccessful()) {
+      final RLPInput rlpInput =
+          RLP.input(privateTransactionSimulatorResultOptional.get().getOutput());
+      if (rlpInput.nextSize() > 0) {
+        return Optional.of(
+            new PrivacyGroup(
+                privacyGroupId,
+                PrivacyGroup.Type.ONCHAIN,
+                "",
+                "",
+                decodeParticipantList(rlpInput.raw())));
+      } else {
```

Comment:
nit: these 2 else returns could be combined

Reply:
Done